src/feeds/FeedReader.ts

```diff
@@ -0,0 +1,144 @@
+import { BridgeConfigFeeds } from "../Config/Config";
+import { ConnectionManager } from "../ConnectionManager";
+import { FeedConnection } from "../Connections";
+import LogWrapper from "../LogWrapper";
+import { MessageQueue } from "../MessageQueue";
+
+import axios from "axios";
+import Parser from "rss-parser";
+
+const log = new LogWrapper("FeedReader");
+
+export class FeedError extends Error {
+    constructor(
+        public url: string,
+        public cause: Error,
+    ) {
+        super(`Error fetching feed ${url}: ${cause.message}`);
+    }
+}
+
+export interface FeedEntry {
+    feed: {
+        title: string,
+        url:   string,
+    },
+    title: string,
+    link:  string,
+}
+
+interface AccountData {
+    [url: string]: string[],
+}
+
+function stripHtml(input: string): string {
+    return input.replace(/<[^>]*?>/g, '');
+}
+
+export class FeedReader {
+    private observedFeedUrls: string[];
+    private seenEntries: Map<string, string[]> = new Map();
+    static readonly seenEntriesEventType = "uk.half-shot.matrix-hookshot.feed.reader.seenEntries";
+
+    constructor(
+        private config: BridgeConfigFeeds,
+        private connectionManager: ConnectionManager,
+        private queue: MessageQueue,
+        private matrixClient: MatrixClient,
+    ) {
+        const feedConnections = this.connectionManager.getAllConnectionsOfType(FeedConnection);
+        this.observedFeedUrls = feedConnections.map(c => c.feedUrl);
+        connectionManager.on('new-connection', c => {
+            if (c instanceof FeedConnection) {
+                log.info('New connection tracked:', c.feedUrl);
+                this.observedFeedUrls.push(c.feedUrl);
+            }
+        });
+
+        log.info('Loaded feed URLs:', this.observedFeedUrls);
+
+        void this.loadSeenEntries().then(() => {
+            return this.pollFeeds();
+        });
+    }
+
+    private async loadSeenEntries(): Promise<void> {
+        const accountData = await this.matrixClient.getAccountData<AccountData>(FeedReader.seenEntriesEventType).catch((err: any) => {
+            if (err.statusCode === 404) {
+                return {} as AccountData;
+            } else {
+                throw err;
+            }
+        });
+        for (const url in accountData) {
+            this.seenEntries.set(url, accountData[url]);
+        }
+    }
+
+    private async saveSeenEntries(): Promise<void> {
+        const accountData: AccountData = {};
+        for (const [url, guids] of this.seenEntries.entries()) {
+            accountData[url.toString()] = guids;
+        }
+        await this.matrixClient.setAccountData(FeedReader.seenEntriesEventType, accountData);
+    }
+
+    private async pollFeeds(): Promise<void> {
+        log.debug(`Checking for updates in ${this.observedFeedUrls.length} RSS/Atom feeds`);
+
+        let seenEntriesChanged = false;
+
+        await Promise.all(this.observedFeedUrls.map(async (url) => {
```

Comment:
Hmm, this feels a bit too thundering herd-y to me. We could end up checking 1000s of feeds at once, and then do nothing for N seconds which would potentially be quite disruptive. Can we stagger them a bit?

Reply:
Adjusted in 420def0 – we now do it sequentially.